### PR TITLE
Add Django 2.2 to test matrix

### PR DESCRIPTION
**Pull request by federicobond**

### CODE REVIEW

**Comment by federicobond** [comment]
I refactored the CI setup to make it harder to skip variants inadvertently. CI is failing some environments due to timing issues in `test_get_last_invalidation` which appear unrelated to these changes.

**Comment by djpretzel** [comment]
@federicobond @noripyt does this effectively mean the current 2.1.0 release works w/ Django 2.2?

**Comment by federicobond** [comment]
Yes, as far as test coverage is concerned. You may suppress the version check via Django settings if you are in a hurry to use Django 2.2.

**Comment by federicobond** [comment]
@BertrandBordage sorry to bother you, would you have some time to take a look at this? Let me know if I can help you in any way.

**Comment by jnoortheen** [comment]
it has been weeks since Django 2.2 is released. There were interesting additions like bulk update. But I am not able to install it because of the version conflicts with pypoetry.

File: cachalot/apps.py [hunk 1]
```diff
@@ -13,7 +13,7 @@
 
 @register(Tags.compatibility)
 def check_django_version(app_configs, **kwargs):
-    if not (1, 11) <= django_version < (2, 2):
+    if not (1, 11) <= django_version < (2, 3):
```
Comment (slurms) [inline cachalot/apps.py]: There will be no Django 2.3 (https://www.djangoproject.com/weblog/2015/jun/25/roadmap/). It'll be 3.0 :).

Libraries should work from 1.11 -> 2.2 with no changes needed. A 3.0 release will require we drop support for 1.11.

**Comment by shacker** [comment]
Very much need to get our project onto Django 2.2 (feature needed). But `pipenv install "django==2.2.3" generates errors because cachalot specifies an upper version boundary for Django at 2.1. Is this PR pretty close?

**Comment by federicobond** [comment]
If you need to upgrade to Django 2.2+ add this to your settings:
```python
SILENCED_SYSTEM_CHECKS = ["cachalot.E003"]
```

I have been using django-cachalot for months now and it works fine with Django 2.2.

**Comment by shacker** [comment]
@federicobond That's a fix one might apply *after* getting Django 2.2.x successfully installed. I am saying that in a pipenv environment, cachalot's requirements literally prevent one from upgrading Django. If you run `pipenv graph` it shows as:

```
django-cachalot==2.1.0
  - Django [required: >=1.11,<2.2,]
```
Because it sets an upper bound for the allowed Django version, `pipenv install "django==2.2.3"` fails, and there is no workaround (pipenv takes these things very seriously!)

**Comment by ritiko** [comment]
Looks like the checks just fail on timeout thresholds. Anyway we can speed this PR review up a little? I'm happy to help if needed. I'd rather not fork the project internally

**Comment by federicobond** [comment]
@shacker ugh, sorry to hear about that. My experience with Pipenv has not been very positive either. A good dependency manager should encourage safety but let people override it if necessary.

**Comment by shacker** [comment]
@federicobond My experience with pipenv has been really great. This incident got me thinking -- the *whole point* of a dependency setting an upper boundary is to say "I will not work if you change that other dependency outside of these boundaries, so don't do it."  By establishing that boundary, cachalot has set a guardrail to keep itself from breaking. Pipenv is doing exactly the right thing by respecting that boundary. 

One could argue that pipenv could provide an option for a forced override. But pipenv really is doing the right thing here, and I'd say that if cachalot has publicly stated in code that it will only work in Django versions lower than 2.2, then that's the rule and it ought to be respected by environment managers. So maybe cachalot could/should consider not being so specific with that in the next release? Don't set an upper boundary if you don't know in advance that it will break in higher Django versions!

**Comment by kptac** [comment]
Can we please get an update on this?

**Comment by talkingtoaj** [comment]
I'm wondering if django-cachalot has been abandoned. No progress on django 2.2 progress for a while

**Comment by shacker** [comment]
It is starting to look that way.  I see that [this fork](https://github.com/LeetCode-OpenSource/django-cachalot) is 4 commits ahead, and seems to have Django 2.2 support. I'm going to try that one.

**Comment by stunaz** [comment]
The library is dead...move on

**Comment by shacker** [comment]
FWIW I just uninstalled this version of cachalot, then installed the fork linked to above, then upgraded to Django 2.2.6 with no issues. Unfortunately that fork does not have an open Issues page, which makes me wonder about its long term plans.

Another alternative I just came across is [django-cacheops](https://github.com/Suor/django-cacheops), which appears to do pretty much the same thing but is actively maintained.

**Comment by fabiocaccamo** [comment]
@noripyt could you consider to move this project to django jazzband?
https://github.com/jazzband

**Comment by BertrandBordage** [comment]
@fabiocaccamo Yes, that’s something I’m considering.
I rarely used this project in the past 4 years, and didn’t use it at all for more than a year, and it’s not likely to change. So if someone has the desire and experience to maintain it from now on, I would be very happy about it :) Otherwise we can put it on Jazzband, yes.

**Comment by talkingtoaj** [comment]
@BertrandBordage Totally understandable. Can I just say 'Kudos' to you for starting this project. It has proven very popular and successful for its set-and-forget nature. Thank you from all of us who have benefitted from your project.

**Comment by fabiocaccamo** [comment]
@BertrandBordage I understand you, don't worry. Thank you again for this great project. 
I would put it on Jazzband as soon as possible.

**Comment by BertrandBordage** [comment]
Thank you for the heartwarming messages :) I worked very hard on this module in 2014-2015, mostly for one project that I don’t work on anymore to cope with django CMS performance issues.
I am trying to start the procedure to move it to Jazzband.
In the meantime, any wannabe maintainer is welcome!

So, I guess I should start by creating an issue on https://github.com/jazzband-roadies/help/issues, right?

**Comment by fabiocaccamo** [comment]
I think that you need to join the organization https://jazzband.co/about#joining then transfer project ownership and pypi permissions.

**Comment by BertrandBordage** [comment]
Ah, you are right! I’m doing that.

**Comment by BertrandBordage** [comment]
See https://github.com/jazzband-roadies/help/issues/172 for the discussion related to moving to Jazzband :)

**Comment by shacker** [comment]
I'll add my voice to the chorus - thanks for creating and supporting this project while you could - it was excellent. 

For anyone who reads this far and saw my note above, I found that django-cacheops to be literally a drop-in replacement, and is actively maintained. I was up and running with it in 10 minutes on python 3.8 and django 2.2.x.

**Comment by abbad** [comment]
Did anybody try Django cache-machine as a replacement? 
https://github.com/django-cache-machine/django-cache-machine

**Comment by Andrew-Chen-Wang** [comment]
Hi all! I'm going to maintain the repository as best I can. @stunaz @talkingtoaj @kptac @mheppner  I'm also taking a look at this PR and test it on local machine. @BertrandBordage Thanks for the awesome work!! I've used this package for a really long time.

**It is very important for everyone to read the comparisons between cachalot, cache-machine, and cacheops. They work very differently, although I highly recommend using cachalot for cold caches and cache-machine for hot caches.** @shacker and @abbad 

Regarding pipenv, I've never personally used it except for at cookiecutter-django, but I didn't set it up. @federicobond 

@fabiocaccamo Unfortunately, the Jazzband people declined to maintain, which is understandable. So I'll do my best to integrate tests into this repo and make sure people can actually install this package

@abbad I use a mix of both cachalot and cache-machine since their counting invalidation resembles what would be used in a hot cache. I typically use cachalot in cold caches (and I'm currently working on a horizontal partition django library that can work well with these two cache libraries) and hot caches with Django-cache-machine.

It appears every test it failing? I'll take a look and I'll probably add pre-commit to this repo as well with some linter checks. Help is definitely welcome @ritiko @slurms

**Comment by shacker** [comment]
> Regarding pipenv..

Though I've moved on from this tool, I want to emphasize that the issues I raised wrt cachalot and pipenv are not *about pipenv* - it just so happens that pipenv is the only tool out there (apparently) that actively listens to what your project tells it. If you specify an upper Django version limit in your dependencies, pipenv is going to take that seriously and refuse to upgrade the whole project! Otherwise what's the point of specifying an upper limit (as has been done in cachalot).

Environment managers other than pipenv could/should also be strict about that. It's just a matter of dependency management specifications being taken literally.

**Comment by Andrew-Chen-Wang** [comment]
@shacker I see. New to the game, so thanks for the input! I did read your above comments and agree that the requirements.txt is making this really difficult (as it definitely mind-boggled me at first when I installed cachalot). That being said, I guess it's time to learn Pipenv and not break the entire package in the process...

**Comment by stunaz** [comment]
@shacker @Andrew-Chen-Wang have you checked https://github.com/python-poetry/poetry

**Comment by Andrew-Chen-Wang** [comment]
This codebase is in #146 with Django 2.2 and 3.0 added.

**Pull request closed without merging**